pricegraph/fuzz/fuzz_targets/orderbook.rs

```diff
@@ -1,9 +1,62 @@
 #![no_main]
+use arbitrary::Arbitrary;
 use libfuzzer_sys::fuzz_target;
-use pricegraph::{Element, Orderbook};
+use pricegraph::{Element, Pricegraph, TokenId, TokenPair};
+use std::iter::once;
+
+// Limit the maximum token id that is allowed to appear in the generated orders. Without this we can
+// make orderbook creation too slow if one order contains a large buy or sell token id.
+const MAX_TOKEN_ID: u16 = 16;
 
 // Fuzz creation and usage of Orderbook.
 
-fuzz_target!(|elements: Vec<Element>| {
-    let _ = Orderbook::from_elements(elements);
+#[derive(Arbitrary, Debug)]
+enum Operation {
+    EstimateExchangeRate {
+        pair: TokenPair,
+        sell_amount: f64,
+    },
+    TransitiveOrderbook {
+        base: TokenId,
+        quote: TokenId,
+        spread: Option<f64>,
+    },
+}
+
+#[derive(Arbitrary, Debug)]
+struct Arguments {
+    elements: Vec<Element>,
+    operation: Operation,
+}
+
+fn largest_token_id(elements: &[Element]) -> Option<TokenId> {
+    elements
+        .iter()
+        .flat_map(|e| once(e.pair.buy).chain(once(e.pair.sell)))
+        .max()
+}
+
+fuzz_target!(|arguments: Arguments| {
+    if largest_token_id(&arguments.elements).unwrap_or(0) > MAX_TOKEN_ID {
+        return;
+    }
+
+    let pricegraph = Pricegraph::new(arguments.elements);
+    match arguments.operation {
```

Comment:
Is this necessary now that `Pricegraph` is immutable? That means that order of operations no longer has any impact?

Reply:
It's not necessary in the sense that the previous operation might make some cases for later operations impossible but I still like this for clarity. And it probably still helps the fuzzer isolate cases.

Reply:
Personally I would find it clearer to have separate fuzz targets for the two methods.

Reply:
Are there any concerns about not being able to fuzz each method individually?

Reply:
> And it probably still helps the fuzzer isolate cases.

**Disclaimer**: I know pretty much nothing about how fuzzers work

This surprises me. Intuitively, I would imagine fuzzers would perform better with a narrower search space.

Reply:
Separate targets are probably even better, yes. What I meant before was doing all the operations in a row. I'll try out the separate targets.

Reply:
> What I meant before was doing all the operations in a row.

Oh yes, in that case I definitely agree.

Reply:
My conclusion for separate targets is that I would like to keep this as a single target because a significant code path is shared in the creation of the orderbook and that we would duplicate the `largest_token_id` filter (of course, could probably put it into a module).
I am going to keep it this way and we can reevalute if we add more fuzzing.